src/main/scala/org/ergoplatform/mining/AutolykosPowScheme.scala

```diff
@@ -0,0 +1,212 @@
+      new ErgoFullBlock(h, blockTransactions, extension, Some(adProofs))
+    }
+  }
+
+  /**
+    * Find a nonce from `minNonce` to `maxNonce`, such that full block created from block candidate `candidateBlock`
+    * will contain correct solution of the Autolykos PoW puzzle.
+    */
+  def proveCandidate(candidateBlock: CandidateBlock,
+                     sk: PrivateKey,
+                     minNonce: Long = Long.MinValue,
+                     maxNonce: Long = Long.MaxValue): Option[ErgoFullBlock] = {
+    proveBlock(candidateBlock.parentOpt,
+      candidateBlock.nBits,
+      candidateBlock.stateRoot,
+      candidateBlock.adProofBytes,
+      candidateBlock.transactions,
+      candidateBlock.timestamp,
+      candidateBlock.extension,
+      sk,
+      minNonce,
+      maxNonce
+    )
+  }
+
+  /**
+    * Check nonces from `startNonce` to `endNonce` for message `m`, secrets `sk` and `x`, difficulty `b`.
+    * Return AutolykosSolution if there is any valid nonce in this interval.
+    */
+  private[mining] def checkNonces(m: Array[Byte], sk: BigInt, x: BigInt, b: BigInt, startNonce: Long, endNonce: Long): Option[AutolykosSolution] = {
+    log.debug(s"Going to check nonces from $startNonce to $endNonce")
+    val p1 = pkToBytes(genPk(sk))
+    val p2 = pkToBytes(genPk(x))
+
+    @tailrec
+    def loop(i: Long): Option[AutolykosSolution] = if (i == endNonce) {
+      None
+    } else {
+      if (i % 1000000 == 0 && i > 0) log.debug(s"$i nonce tested")
+      val nonce = Longs.toByteArray(i)
+      val d = (x * genIndexes(Bytes.concat(m, nonce)).map(i => genElement(m, p1, p2, Ints.toByteArray(i))).sum - sk).mod(q)
+      if (d <= b) {
+        log.debug(s"Solution found at $i")
+        Some(AutolykosSolution(genPk(sk), genPk(x), nonce, d))
+      } else {
+        loop(i + 1)
+      }
+    }
+
+    loop(startNonce)
+  }
+
+  /**
+    * Get target `b` from encoded difficulty `nBits`
+    */
+  private[mining] def getB(nBits: Long): BigInt = {
+    q / RequiredDifficulty.decodeCompactBits(nBits)
+  }
+
+  /**
+    * Calculate header fields based on parent header
+    */
+  protected def derivedHeaderFields(parentOpt: Option[Header]): (ModifierId, Byte, Seq[ModifierId], Int) = {
+    val interlinks: Seq[ModifierId] =
+      parentOpt.map(parent => new PoPoWProofUtils(this).constructInterlinkVector(parent)).getOrElse(Seq.empty)
+
+    val height = parentOpt.map(parent => parent.height + 1).getOrElse(ErgoHistory.GenesisHeight)
+
+    val version = Header.CurrentVersion
+
+    val parentId: ModifierId = parentOpt.map(_.id).getOrElse(Header.GenesisParentId)
+
+    (parentId, version, interlinks, height)
+  }
+
+  /**
+    * Hash function that takes `m` and `nonceBytes` and returns a list of size `k` with numbers in
+    * [0,`N`)
+    */
+  private def genIndexes(seed: Array[Byte]): Seq[Int] = {
+    val hash = Blake2b512(Bytes.concat(seed))
+    hash.grouped(3).take(k).toSeq.map(b => Math.abs(Ints.fromByteArray(0.toByte +: b) % N))
```

Comment:
1. Math.abs is not needed, your integer is always positive by construction
2. 3 bytes is about 16.7M, so N = 10^8 = 100M does not make any sense.
3. There is no need to use Blake2b512. If you need for 512 bits, you can use two calls with different prefixes (e.g. 0 and 1), it will be about similar performance, but no additional primitive needed. But I think it could be enough to use 256 bits, so bytes (0-2) for the first index, bytes (1-3) for the second etc.